Mark Huffman Decoder Assembly `noexecstack` on All Architectures

felixhandte opened the pull request:
Apparently, even when the assembly file is empty (because `ZSTD_ENABLE_ASM_X86_64_BMI2` is false), it still is marked as possibly needing an executable stack and so the whole library is marked as such. This commit applies a simple patch for this problem by moving the noexecstack indication outside the macro guard.

In addition, this PR ships flags to the Makefile-driven builds to explicitly request a non-executable stack. This ought to be redundant, but it appears on at least one platform (MIPS) to succeed where the source code note fails.

This PR builds on #2857.

This PR addresses #2963.

felixhandte committed:
Mark Huffman Decoder Assembly `noexecstack` on All Architectures

Apparently, even when the assembly file is empty (because
`ZSTD_ENABLE_ASM_X86_64_BMI2` is false), it still is marked as possibly
needing an executable stack and so the whole library is marked as such. This
commit applies a simple patch for this problem by moving the noexecstack
indication outside the macro guard.

This commit builds on #2857.

This commit addresses #2963.

felixhandte committed:
Add Test Validating Stack is not Executable in playTests.sh

terrelln commented on lib/decompress/huf_decompress_amd64.S:
```diff
@@ -1,14 +1,14 @@
  * ref: https://wiki.gentoo.org/wiki/Hardened/GNU_stack_quickstart
  */
 #if defined(__linux__) && defined(__ELF__)
 .section .note.GNU-stack,"",%progbits
```
We may also want a `&& defined(__GNUC__)`

felixhandte committed:
Makefiles: Add `noexecstack` Options to Compilation and Linking

Hopefully this marks the binary artifacts `noexecstack` even on platforms
where binaries default to true.

felixhandte committed:
Clean Up Debugging Statements

felixhandte committed:
Write `GNU-stack` Section on All ELF Architectures

Previously we did this only on Linux, which missed other Unices.

felixhandte committed:
Restrict `GNU-stack` Note to GNU Assemblers
